rse-py-package.Rmd

```diff
@@ -0,0 +1,1477 @@
+
+You can also use the command line tool `check_zipf.py` which tests to see if 
+a provided list of countsfollows a Zipf distribution::
+
+    $ check_zipf.py 100 50 33 25
+    True: [100, 50, 33, 25]
+
+Authors
+-------
+
+Terry Pratchett
+```
+
+In addition to the README, we should also make sure that there are
+[docstrings](#docs-docstrings) written for the package as a whole
+(in the `__init__.py`):
+
+<!-- src="rse-py-package/09/zipfpy/__init__.py" -->
+```python
+"""
+The zipfpy package contains tests of distributions of counts to see if
+they follow a Zipf distribution (https://en.wikipedia.org/wiki/Zipf%27s_law)
+using the routines in the check module, and routines for generating
+lists of counts that follow a Zipf distribution in the generate module.
+"""
+RELATIVE_ERROR = 0.05
+```
+
+and for key routines like `is_zipf` and `make_zipf`:
+
+<!-- src="rse-py-package/09/zipfpy/check.py" -->
+```python
+def is_zipf(hist, rel=RELATIVE_ERROR):
+    """Tests if a histogram of counts follows a Zipf distribution.
+
+    Given a list of counts as hist, assumed sorted in decreasing order,
+    and a relative error tolerance (if not provided, the default value
+    zipfpy.RELATIVE_ERROR is used), tests to see if the counts follow
+    a Zipf distribution.
+
+    Args:
+        hist: an list or other iterable containing a list of numeric counts
+        rel: the relative error tolerance used if provided; if not,
+             the package default is used.
+
+    Returns:
+        True if the list of counts follows a zipf distribution within
+        the relative tolerance.  False otherwise.
+
+    Raises:
+        AssertionError: raised if an empty list is passed.
+    """
+    assert len(hist) > 0, 'Cannot test Zipfiness without data'
+    scaled = [h/hist[0] for h in hist]
+    perfect = make_zipf(len(hist))
+    return scaled == approx(perfect, rel=rel)
+```
+
+<!-- src="rse-py-package/09/zipfpy/generate.py" -->
+```python
+def make_zipf(length):
+    """Returns a list of counts that follows a Zipf distribution.
+
+    Args:
+        length: the number of counts to be generated
+
+    Returns:
+        A list of the provided length of floating point numbers corresponding
+        exactly the zipf distribution.  For example, for length=5:
+
+        [1.0, 0.5, 0.3333333333333333, 0.25, 0.2]
+
+    Raises:
+        AssertionError: raised if a zero or negative length is provided
+    """
+    assert length > 0, 'Zipf distribution must have at least one element'
+    result = [1/(1 + i) for i in range(length)]
+    return result
+```
+
+Not only is the documentation then useful for those reading the 
+code, but it helpfully shows up when a user types, for instance,
+`help(zipfpy)`.
+
+<!-- used="rse-py-package/09/requirements.txt" -->
+<!-- used="rse-py-package/09/setup.py" -->
+<!-- used="rse-py-package/09/test_zipf.py" -->
+
+## How do I generate web pages of documentation for my package? {#rse-py-package-sphinx}
```

Comment:
Although ReST is the canonical Python documentation markup, I think it could be somewhat daunting to feel like you need to learn a new language to write documentation. It might therefore be worthwhile mentioning that sphinx works with markdown documents even if it doesn't support directives and roles to the same extent (I believe), to lower the threshold for getting started, especially for simpler projects markdown can be completely sufficient and is also supported at RTD.

- https://stackoverflow.com/questions/2471804/using-sphinx-with-markdown-instead-of-rst
- https://www.sphinx-doc.org/en/master/usage/markdown.html
- https://docs.readthedocs.io/en/latest/intro/getting-started-with-sphinx.html#using-markdown-with-sphinx

There is also [MKdocs](https://docs.readthedocs.io/en/latest/intro/getting-started-with-mkdocs.html), but now that sphinx supports markdown I don't feel like it is necessary to go through, not sure if it is worth briefly mentioning (although I am not an expert on the differences versus sphinx with markdown).